etcd3/baseclient.py

```diff
@@ -48,15 +52,59 @@ def __iter__(self):
 DEFAULT_VERSION = '3.3.0'
 
 
+def retry_all_hosts(func):
+    def wrapper(self, *args, **kwargs):
+        errors = []
+        got_result = False
+        call_endpoints = copy.copy(self.endpoints)
```

Comment:
maybe a lock is better than copy

if we have 3 endpoints of which the initial 2 are invalid:
```
client.endpoints:      [invalid, invalid, valid]
and after 1 request:
without copy   ->      [valid, invalid, invalid]       the next request will access the valid endpoint directly   
with copy      ->      [invalid, invalid, valid]       the next request still have to try 3 times
```

Reply:
There are two problems to consider:
1) the endpoint whose response is accepted should become the active one
2) the api call should use the endpoints defined at the start of the call (to prevent problems with concurrent code overwriting `.endpoints`)

A lock could work, but to prevent `2` it should lock the whole `while retries > 0` loop and so prevent concurrent calls.

It should work correctly if when the while loop operating on the copy exits correctly, I add a block:
`````
if call_endpoints[0] in endpoints:
  # move around endpoints to have endpoints[0] == call_enpoints[0]
else:
  # .endpoints has been modified externally
  # leave it as it is, the next call will find a good one
`````

Reply:
no need to lock the whole `while retries > 0`

in fact, atomicity seems not so important here

or, just lock `self.endpoints = xxxxx`